projects/RabbitMQ.Client/client/ClientMemory.cs

```diff
@@ -0,0 +1,88 @@
+﻿// This source code is dual-licensed under the Apache License, version
+// 2.0, and the Mozilla Public License, version 2.0.
+//
+// The APL v2.0:
+//
+//---------------------------------------------------------------------------
+//   Copyright (c) 2007-2020 VMware, Inc.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       https://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//---------------------------------------------------------------------------
+//
+// The MPL v2.0:
+//
+//---------------------------------------------------------------------------
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+//  Copyright (c) 2007-2020 VMware, Inc.  All rights reserved.
+//---------------------------------------------------------------------------
+
+using System;
+
+namespace RabbitMQ.Client
+{
+    public struct ClientMemory : IDisposable
```

Comment:
I don't get the value add of this structure. There is already such a type built into the framework https://learn.microsoft.com/en-us/dotnet/api/system.buffers.memoryhandle. You can optain those by switching to memory pool. But I wonder why this is needed. What's the benefit if having to expose the pooling details to consumers? I saw a comment on basic get that argued about removing disposable there but now this type is also public and disposable so effectively things have only be moved.

From an API perspective I think you should stick to either ReadonlyMemory or BinaryData and then treat it as either externally or internally managed depending on the use case and add necessary copying where needed. For me having browsed through these changes I have my doubts about this types valueadd

Reply:
Great, I'm unaware of that type.

I did not originally add the use of `ArrayPool` and perhaps at that time `MemoryPool` was not available.

Since this PR has gone on long enough I will add a separate issue to look into this further. Thanks!

Reply:
`MemoryPool` may not be an option. At any rate, let's continue discussion here - https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1421

Reply:
@danielmarbach FYI in https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1347/commits/1fca1ba9ec472667924a3360857eedba461d2454 I reverted to using `ReadOnlyMemory<byte>` for the API because, as you pointed out, there is _no reason at all_ to expose the internal structure. Duh. Thank you for pointing that out!